src/backend/distributed/transaction/lock_graph.c

```diff
@@ -0,0 +1,772 @@
+	 */
+
+	/* build list of starting procs */
+	for (curBackend = 0; curBackend < MaxBackends; curBackend++)
+	{
+		PGPROC *currentProc = &ProcGlobal->allProcs[curBackend];
+		BackendData *currentBackendData = NULL;
+
+		/* skip if the PGPROC slot is unused */
+		if (currentProc->pid == 0)
+		{
+			continue;
+		}
+
+		currentBackendData = GetBackendDataForProc(currentProc);
+
+		/*
+		 * Only start searching from distributed transactions originating on the source
+		 * node. Other deadlocks may exist, but the source node can only resolve those
+		 * that involve its own transactions.
+		 */
+		if (sourceNodeId != currentBackendData->transactionId.initiatorNodeIdentifier ||
+			!IsInDistributedTransaction(currentBackendData))
+		{
+			continue;
+		}
+
+		/* skip if the process is not blocked */
+		if (!IsProcessWaitingForLock(currentProc))
+		{
+			continue;
+		}
+
+		remaining.procs[remaining.procCount++] = currentProc;
+	}
+
+	while (remaining.procCount > 0)
+	{
+		PGPROC *waitingProc = remaining.procs[--remaining.procCount];
+
+		/*
+		 * We might find a process again if multiple distributed transactions are
+		 * waiting for it, but this won't result in any new edges. Moreover, if there
+		 * is a local deadlock we might keep going forever. Therefore we skip processes
+		 * that we've already visited.
+		 */
+		if (visited[waitingProc->pgprocno])
+		{
+			continue;
+		}
+
+		visited[waitingProc->pgprocno] = true;
+
+		/* only blocked processes result in wait edges */
+		if (!IsProcessWaitingForLock(waitingProc))
+		{
+			continue;
+		}
+
+		/*
+		 * Record an edge for everyone already holding the lock in a
+		 * conflicting manner ("hard edges" in postgres parlance).
+		 */
+		AddEdgesForLockWaits(waitGraph, waitingProc, &remaining);
+
+		/*
+		 * Record an edge for everyone in front of us in the wait-queue
+		 * for the lock ("soft edges" in postgres parlance).
+		 */
+		AddEdgesForWaitQueue(waitGraph, waitingProc, &remaining);
+	}
+
+	UnlockLockData();
+
+	return waitGraph;
+}
+
+
+/*
+ * LockLockData takes locks the shared lock data structure, which prevents
+ * concurrent lock acquisitions/releases.
+ */
+static void
+LockLockData(void)
+{
+	int partitionNum = 0;
+
+	LockDistributedTransactionSharedMemory(LW_SHARED);
+
+	for (partitionNum = 0; partitionNum < NUM_LOCK_PARTITIONS; partitionNum++)
+	{
+		LWLockAcquire(LockHashPartitionLockByIndex(partitionNum), LW_SHARED);
+	}
+}
+
+
+/*
+ * UnlockLockData unlocks the locks on the shared lock data structure in reverse
+ * order since LWLockRelease searches the given lock from the end of the
+ * held_lwlocks array.
+ */
+static void
+UnlockLockData(void)
+{
+	int partitionNum = 0;
+
+	for (partitionNum = NUM_LOCK_PARTITIONS - 1; partitionNum >= 0; partitionNum--)
+	{
+		LWLockRelease(LockHashPartitionLockByIndex(partitionNum));
+	}
+
+	UnlockDistributedTransactionSharedMemory();
+}
+
+
+/*
+ * AddEdgesForLockWaits adds an edge to the wait graph for every lock that
```

Comment:
>  AddEdgesForLockWaits adds an edge to the wait graph for every lock that waitingProc is waiting for.

I still think that this comment needs some clarification. We're only iterating through `waitingProc->waitLock-> procLocks`. So isn't it `Adds an edge to the wait graph for every process that holds the lock (that waitingProc is waiting on) in conflicting manner`

Or something with better words?